### PR TITLE
KMeans: Don't try to normalize sparse data; show warning

### LINKED ISSUE #4973: k-Means: normalization fails when mixing original and bow attributes
**Bug reports that do not sufficiently describe a bug are not helpful and will most likely be ignored.**

- [ ] **A clear and concise description.** "*I'm getting an error when I open a tree viewer*" is far from enough (what kind of error? what kind of data you give it?). "*I'm getting an error*" is even worse.

- [ ] **How to reproduce the problem.** Upload a zip with a workflow and data, or describe the steps (open this widget, click there, then add this ...)

- [ ] **Which version of Orange are you using? How did you install it?**

- [ ] **What is your operating system?**

Orange Version: 3.26.0
Operational System: macOS Mojave Version 10.14.6 - Mac Pro Mid 2010

The same file is working in my other Macbook Pro. 

How can I uninstall Orange completely to reinstall it again? Is there any way to do it? Because just removing and reinstalling the Orange maintain it's structures.

ERROR MESSAGE:

Error encountered in widget k-Means:
```
Traceback (most recent call last):
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/orangecanvas/scheme/signalmanager.py", line 936, in __process_next
    if self.__process_next_helper(use_max_active=True):
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/orangecanvas/scheme/signalmanager.py", line 974, in __process_next_helper
    self.process_node(selected_node)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/orangecanvas/scheme/signalmanager.py", line 605, in process_node
    self.send_to_node(node, signals_in)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/orangewidget/workflow/widgetsscheme.py", line 792, in send_to_node
    self.process_signals_for_widget(node, widget, signals)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/orangewidget/workflow/widgetsscheme.py", line 826, in process_signals_for_widget
    handler(*args)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/widgets/utils/sql.py", line 31, in new_f
    return f(widget, data, *args, **kwargs)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/widgets/unsupervised/owkmeans.py", line 578, in set_data
    self.invalidate(unconditional=True)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/widgets/unsupervised/owkmeans.py", line 458, in invalidate
    self.unconditional_commit()
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/widgets/unsupervised/owkmeans.py", line 444, in commit
    self.run_optimization()
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/widgets/unsupervised/owkmeans.py", line 403, in run_optimization
    self.__launch_tasks(needed_ks)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/widgets/unsupervised/owkmeans.py", line 357, in __launch_tasks
    preprocessed_data = self.preproces(self.data)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/widgets/unsupervised/owkmeans.py", line 493, in preproces
    data = Normalize()(data)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/preprocess/preprocess.py", line 356, in __call__
    return normalizer(data)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/preprocess/normalize.py", line 36, in __call__
    return data.transform(domain)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/data/table.py", line 520, in transform
    return type(self).from_table(domain, self)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/orangecontrib/text/corpus.py", line 508, in from_table
    t = super().from_table(domain, source, row_indices)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/data/table.py", line 463, in from_table
    variables=domain.attributes)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/data/table.py", line 397, in get_columns
    col_array = match_density(col(source))
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/preprocess/transformation.py", line 38, in __call__
    transformed = self.transform(col)
  File "/Applications/Orange3.app/Contents/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/Orange/preprocess/transformation.py", line 127, in transform
    raise ValueError('Normalization does not work for sparse data.')
ValueError: Normalization does not work for sparse data.
```

Thank you .

### PR DESCRIPTION
**Pull request by janezd**
##### Issue

Fixes #4973.

##### Description of changes

K-means fails when it tries to normalize sparse data. Normalization of sparse data is impossible (because it turns it dense). Hence the widget now doesn't normalize but shows a warning instead.

##### Includes
- [X] Code changes
- [X] Tests

### CODE REVIEW

File: Orange/widgets/unsupervised/owkmeans.py [hunk 1]
```diff
@@ -130,6 +132,7 @@ class Warning(widget.OWWidget.Warning):
         not_enough_data = widget.Msg(
             "Too few ({}) unique data instances for {} clusters"
         )
+        no_sparse_normalization = widget.Msg("Sparse datacannot be normalized")
```
Comment (ajdapretnar) [inline Orange/widgets/unsupervised/owkmeans.py:135]: Sparse datacannot be normalized --> Sparse data cannot be normalized

Comment (janezd) [reply Orange/widgets/unsupervised/owkmeans.py]: Oops. Fixed.

**Comment by ajdapretnar** [comment]
I like this and I can merge, but wouldn't it make sense to disable normalize checkbox when the data is sparse?

**Comment by janezd** [comment]
> I like this and I can merge, but wouldn't it make sense to disable normalize checkbox when the data is sparse?

This would be even easier to implement. However, imagine a user who leaves this check box checked, always normalizing the data. When she inputs some sparse data (perhaps even without realising it, maybe because she's not even aware of the difference between dense and sparse), the checkbox would be suddenly disabled (and correctly showed as unchecked, of course).

Isn't it better (= less surprising, less behind the back) to leave it and tell the user "Hey, you know I'm not normalizing this, right?"

**Comment by ajdapretnar** [comment]
I meant AND not OR. :D Meaning I would disable it, too. And trigger a warning. For those that don't bother reading warnings (as we have noticed). And also if one is including a screenshot somewhere, it is more transparent IMO.

**Comment by janezd** [comment]
I added a commit that implements "AND". But it is a bit weird.

Checkboxes that are disabled can define `stateWhenDisabled`. If I set it to `False`, the checkbox would be shown as disabled. In this case, showing a warning wouldn't be OK, because normalization looks as if it's disabled (it only WAS enabled before the checkbox was disabled).

Hence I haven't set the state. The checkbox is shown as checked, but disabled and a warning is shown. What the user sees is now consistent, but she can't do anything to get rid of the warning because the checkbox is disabled!

If however, the checkbox was unchecked before the widget got the data, it is only disabled and the warning is not shown. This is OK.

All in all, I'd remove this last commit, but I guess I won't argue if you decide it should stay.

**Comment by ajdapretnar** [comment]
I see your point. I will leave it up to you, because I agree there is no good solution, even though I am slightly more in favor of disabling controls and keeping the warning (can't remove it? no harm done). But it is not vital - important this is, the problem is solved.

**Comment by janezd** [comment]
If you prefer with disabled check box, merge it as is.

**Pull request merged**